CI docs (#182)
ignitionrobotics/docs: ci.md

# Ignition Continuous Integration

When developers push code changes into Ignition, a series of automated tests
run for various platforms. This process is called Continuous Integration, or CI.
This document covers all the CI jobs that are available for Ignition and how to
interpret their results.

## Types of checks

The main type of check that is performed by CI is compiling the code and running
all automated tests. Some jobs may also run linters and other static checkers.
Finally, we have checks that add labels, check DCO and coverage.

All checks are visible at the bottom of a pull request.

![PR checks](images/PR_checks.png)

We use two platforms to run CI, each of them running different builds.

### Jenkins server

Our Jenkins server is located at https://build.osrfoundation.org/. It runs builds

> for all supported operating systems. It also runs an ABI checker for all stable

Reviewer: Should we define "stable," since it's referred to so many times? Does stable == release?

Author: Good point, I'll link to these docs: https://github.com/ignitionrobotics/docs/blob/master/release.md#type-of-releases

branches.

Each pull request to a stable or `main` branch will trigger:

* `<library>-ci-pr_any-ubuntu-auto-amd64`: Build and test on Ubuntu Linux
    * Builds are run inside Docker containers for Bionic or Focal.
    * By default, dependencies are installed from stable debian binaries for
      stable branches, and nightlies for `main` branches.
    * Library being tested is compiled using CMake and Make.
* `<library>-ci-pr_any-homebrew-amd64`: Build and test on macOS
    * Dependencies are installed from homebrew.
    * Library being tested is compiled using CMake and Make.
* `<library>-ci-pr_any-windows*-amd64`: Build and test on Windows
    * External dependencies are installed from vcpkg

>     * Library being tested and its Ignition dependencies are compiled using colcon.

Reviewer 2: ```suggestion
    * Library being tested and its Ignition dependencies are compiled using [colcon](https://colcon.readthedocs.io/en/released/#).
```

Pull requests to stable branches also trigger:

* `<library>-abichecker-any_to_any-ubuntu_auto-amd64`: Run ABI checker

> ### GitHub Actions

Reviewer 2: It would be nice to include a summary/table at the end of this section that lists what checks are different between Jenkins and GitHub actions.

Author: I turned the entire section into a table :smirk:

GitHub actions run on GitHub's servers. Currently, it builds Linux for all
libraries, and macOS for some of them.

Every commit to any branch will trigger:

* `Ubuntu CI / Ubuntu Bionic CI (push)`: Build, test and run linters on Ubuntu Bionic.
    * Builds are run inside Docker containers.
    * By default, dependencies are installed from stable debian binaries for
      stable branches, and nightlies for `main` branches.
    * Library being tested is compiled using CMake and Make.

>     * Test coverage is computed and uploaded to Codecov.

Reviewer: Is test coverage only for Bionic and not Focal?
The Bionic and Focal items are exactly the same other than the distro name. It felt dizzy reading this dense technical text twice. Is it possible to combine them?

Author: Yeah the only difference is that we only run coverage for one of them, because we can't keep track of multiple coverage results for the same branch. I agree that combining should make it easier to read :+1:

* `Ubuntu CI / Ubuntu Focal CI (push)`: Build, test and run linters on Ubuntu Focal.
    * Builds are run inside Docker containers.
    * By default, dependencies are installed from stable debian binaries for
      stable branches, and nightlies for `main` branches.
    * Library being tested is compiled using CMake and Make.

Every pull request will also trigger the exact same builds, but with the
`pull_request` suffix. Note that pull requests from forks won't have the
`push` checks, while pull requests from the official repositories will have

> both `pull_request` and `push`.

Reviewer 2: I think this is the first time we mention `push` checks - this is just a check for a commit, right? Maybe we should mention that here, or in line 49 where we first discuss what commits trigger.

> Other checks run on GitHub Actions:

Reviewer 2: I find it a little awkward that just before this, we discuss bionic vs focal checks for GitHub Actions, but then here, we are discussing other checks which seem to be run on both bionic and focal. As stated in https://github.com/ignitionrobotics/docs/pull/182/files#r627050201, it seems like the bionic vs focal checks section has some duplicate information, so maybe we can merge that with the comments in this section for an easier read.

* `DCO`: Checks that all commits are signed correctly for the

>     [Developer Certificate of Origin](https://developercertificate.org/).

Reviewer: I don't know if we have information on DCO for beginner contributors, but I was very confused in the beginning, when GitHub gave me a command to run to fix DCO, and it ended up signing alllllll the commits, even ones I merged in from other people. To fix the fix, I basically rechecked out the repo... There are more helpful commands online than the one GitHub gives.

Author: We have some guidance in the contribution guide, so I didn't want to repeat that information here. I'll add a link: https://ignitionrobotics.org/docs/all/contributing#process

* `PR Collection Labeler`: Adds collection labels (i.e. Blueprint, Citadel...)
    to the pull request automatically according to the target branch.
* `Ticket opened`: Adds the pull request to the [Core development board](https://github.com/orgs/ignitionrobotics/projects/3).
* `codecov/*`: Checks that the test coverage hasn't been reduced.

## Required checks

Some checks are marked as `Required`. This means the pull request can't
be merged unless that check is green.

Ideally, all checks would be required. But known warnings, test failures and
flaky tests often prevent us from marking them as required. Until these are
addressed, some checks will need to be checked by hand when they fail, to see
if those failures are unexpected. See

> [this issue](https://github.com/ignition-tooling/release-tools/issues/398) for

Reviewer: Could there be a one-sentence description in that issue saying, the checked things are required? That wasn't obvious from the title, with "check" meaning different things in "CI checks" and "All checked" in the checkbox lists. A bit of a mind pause there.

Author: I added a few sentences :sweat_smile: Once this guide is live, we can link to it from that issue for better context.

what checks are required for each library.

## Interpreting results

On the GitHub UI, checks can be:

> * ✅: Passed, all good!

Reviewer: ✅ should be ✔️
Just kidding..........

Reviewer: Ohhh interesting. On Ubuntu Chrome, the former is white and latter green. On Android, the former is green and latter black 🤷‍♀️

Author: I was annoyed by the same thing :sweat_smile: Gotta ask the Unicode Consortium to be more specific about check colors, those render differently according to the platform :grimacing: 

![image](https://user-images.githubusercontent.com/5751272/117331916-1354bc80-ae4c-11eb-8c3c-20499aa13f92.png)

* 🟡: Pending, results haven't been received yet.
* ❌: Failed, something is wrong.

> Depending on the library and on the build queue, checks can be in a pending

Reviewer 3: If I just started and had no experience with CI, I don't think I would have known where the build queue was

```suggestion
Depending on the library and on the build queue (from our [Jenkins server](https://build.osrfoundation.org/)), checks can be in a pending
```

  before merging the pull request.
* If the build is not required, investigation is needed to find out if the
  failure is pre-existing or if it is being introduced by the pull request.

Clicking on the `Details` button of a failing build takes you to the build so
that you can inspect it.

Builds can fail for a variety of reasons, for example:

* The code failed to compile. This should never be accepted, even in non-required
  checks, and must be fixed.
    * Jenkins: When there's a compilation failure, the build is marked red 🔴.
               On the build's page, click on `Console Output` then scroll until
               you find the compilation errors under the `compiling` section.
    * Actions: When there's a compilation failure, the build is marked red ❌.
               On the build logs, the compilation failure should be under the
               `make` collapsible.
* There are test failures. Test failures that aren't pre-existing must be fixed.
    * Jenkins: When there are test failures, the build is marked yellow 🟡.
               On the build's page, the failing tests are listed under
               `Test Result`. You can also check the `Console Output` for the

>                full test logs.

Reviewer: Should we add a note about how to get to the History page of the test? I found that tip to be very helpful in telling me exactly where to look to determine whether it's a new failure or preexisting. If I remember correctly, there were two or more places that say "History" you can click on, and one of them would give me the graph I want. The page seemed different for different test sometimes, and I don't know why that should be.

    * Actions: When there are test failures, the build is marked red ❌.
               On the build logs, the test failures should be under the
               `make test` collapsible.
    * Check which tests are known failures by searching for the test name on
      issues.
* There are warnings. Warnings that aren't pre-existing must be fixed.
    * Jenkins: When there are warnings, the build is marked yellow 🟡.
               On the build's page, the warnings are listed under
               `GNU C Compiler` and `CMake`. You can also check the
               `Console Output` for the full warning logs.
    * Actions: Does not detect compiler or CMake warnings.
* There are static checker failures. All code checker failures must be fixed.
    * Jenkins: Does not run static checkers.

>     * Actions: When there are code cherk failures, the build is marked red ❌.

Reviewer 3: ```suggestion
    * Actions: When there are code check failures, the build is marked red ❌.
```

               On the build logs, the checker failures should be under the
               `Code check` collapsible.

> * There are infrastructure failures. These should be reported to the build farmer.

Reviewer 2: How do external contributors report to the build farmer? Should they create/add to an issue? If so, where can this issue be created/found?

> * There are infrastructure failures. These should be reported to the build farmer.

Reviewer 3: Might be good to add a short statement on who the build farmer is and what they do? :farmer:

opened. But often it's necessary to re-trigger CI without making changes to the
code if:

* There were infrastructure failures in a previous build.
* There are fixes upstream, on infra or dependencies, that may make a new build pass.

The following methods can be used to re-trigger builds:

* Jenkins
    * Make a comment starting with `@osrf-build run tests` on the pull request
      and all Jenkins builds will be re-triggered.
    * To restart just one specific build and avoid re-running builds that aren't
      necessary, go to the failing build and click `Retry`. This is only available
      to maintainers.
* Actions
    * On the top-right of a build, click `Re-run jobs`. This button is only
      available to maintainers and sometimes mysteriously disappears.

### Custom branches

Maintainers can manually trigger builds for custom branches without opening pull
requests, optionally building dependencies from source and changing other
configurations for testing.

* Actions: Builds are run for every commit. For building dependencies from source,
           add a `.github/ci/dependencies.yaml` file as described

>            [here](https://github.com/ignition-tooling/action-ignition-ci#source-dependencies).

Reviewer: I think I'm just not understanding this because I've never had to do this. Is this for if you are adding things that will require new dependencies, and this adds dependencies to CI so that you can test that it works, before officially adding a dependency to the repo?
Is there an example for this bullet? I think the example in the next bullet is helpful (though I've never had to do that either, so can't comment on that).

Author: I'll add more information about when this may be needed. It's tough to be concise and thorough at the same time :sweat_smile: 

One possible reason to use this is testing changes across repos. If your `ign-gazebo` PR needs a change from `ign-physics`, you could build `ign-physics` from source to check it works on CI before even opening the `ign-physics` PR, or making a release.

Reviewer: Ah okay. Then I've actually come across the need for that but wasn't aware how it can be done. I think saying that scenario helps.

Reviewer 2: On a related note, I just looked at this link, and I think it'd be nice to clarify what to do for `Be sure to add the apt dependencies of dependencies build from source to apt-dependencies`, because I honestly have no clue how to do this after reading that sentence :upside_down_face:

Reviewer 3: If we later want to open a PR, would we have to remove the dependency file?

Author: Ok, this _should_ make things clearer: https://github.com/ignition-tooling/action-ignition-ci/pull/37

Our Jenkins server is located at
[https://build.osrfoundation.org/](https://build.osrfoundation.org/).
It runs builds for all supported operating systems. It also runs an
[ABI](https://stackoverflow.com/a/2456882/6451468)
checker for all stable branches.

GitHub actions run on GitHub's servers. Currently, it builds Linux for all
libraries, and macOS for some of them.

Here's a summary of all checks.

Platform | Job name | OS | What does it do? | Where dependencies come from?
--- | --- | --- | --- | ---
Jenkins | `<library>-ci-pr_any-ubuntu-auto-amd64` | Ubuntu | Compile and run tests using CMake and Make | Stable binaries for stable branches, nightlies for `main`
Jenkins | `<library>-ci-pr_any-homebrew-amd64` | macOS | Compile and run tests using CMake and Make | Homebrew binaries
Jenkins | `<library>-ci-pr_any-windows*-amd64` | Windows | Compile and run tests using [colcon](https://colcon.readthedocs.io/en/released/#) | External dependencies from [vcpkg](https://docs.microsoft.com/en-us/cpp/build/vcpkg?view=msvc-160), Ignition dependencies built from source with colcon
Jenkins | `<library>-abichecker-any_to_any-ubuntu_auto-amd64` | Ubuntu | Run ABI checker | Stable binaries

> Actions | `Ubuntu CI / Ubuntu * CI` | Ubuntu Bionic and Focal | Compile and run tests using CMake and Make, run code checker and upload coverage results | Stable binaries for stable branches, nightlies for `main`

Reviewer 2: This is a nice chart! Should we mention why we are running tests on both GitHub actions `Ubuntu CI` and Jenkins `<library>-ci-pr_any-ubuntu-auto-amd64`? Readers might wonder why Ubuntu tests are being done twice (and to be honest, I've always wondered this myself as well :upside_down_face:)

Author: Added in 191a329

                   left to see if the test has failed before.
                   It's often helpful to see the full test logs on `Console Output`.
    * **Actions**: When there are test failures, the build is marked red ❌.
                   On the build logs, the test failures should be under the
                   `make test` collapsible.
    * Check which tests are known failures by searching for the test name on
      issues.
* There are **warnings**. Warnings that aren't pre-existing must be fixed.
    * **Jenkins**: When there are warnings, the build is marked yellow 🟡.
                   On the build's page, the warnings are listed under
                   `GNU C Compiler` and `CMake`. You can also check the
                   `Console Output` for the full warning logs.
    * **Actions**: Does not detect compiler or CMake warnings.
* There are **static checker failures**. All code checker failures must be fixed.
    * **Jenkins**: Does not run static checkers.
    * **Actions**: When there are code check failures, the build is marked red ❌.
                   On the build logs, the checker failures should be under the
                   `Code check` collapsible.

> * There are **infrastructure failures**. Contributors should leave a comment to
>     maintainers, who willl report these to the build farmer.

Reviewer 2: ```suggestion
* There are **infrastructure failures**. Contributors should leave a comment to
    maintainers (this can be done by leaving a comment on the PR), who will report these to the build farmer.
```

requests, optionally building dependencies from source and changing other
configurations for testing.

One possible reason to trigger CI with custom dependency branches is testing
changes across repositories. For example, if your `ign-gazebo` pull request
needs a change from `ign-physics`, you could build a custom branch of `ign-physics`
from source to check it works with `ign-gazebo` on CI before even opening an
`ign-physics` pull request, or making a release.

#### Actions

By default, builds are run for every commit based on configuration inside the
`.github/ci` directory. So it's possible to change the contents of that
directory and immediately see updated CI results when you push commits.

See the documentation on
[action-ignition-ci](https://github.com/ignition-tooling/action-ignition-ci)
to see what can be configured and how.

Be sure to revert these changes before merging the pull request.

#### Jenkins

The `pr_any` jobs can be triggered manually for any branch by maintainers.
Go to the job's page, for example
[ignition_launch-ci-pr_any-ubuntu_auto-amd64](https://build.osrfoundation.org/job/ignition_launch-ci-pr_any-ubuntu_auto-amd64/),

> click on `Build with Parameters` to trigger a new build.

Reviewer 4: I would leave a warning that this sometimes causes problems and link to https://github.com/ignition-tooling/release-tools/issues/242

    * **Jenkins**: Does not run static checkers.
    * **Actions**: When there are code check failures, the build is marked red ❌.
      On the build logs, the checker failures should be under the `Code check`
      collapsible.
* There are **infrastructure failures**. Contributors should leave a comment on
    the pull request for maintainers, who will report these to the build farmer.
    * **Jenkins**: When there's an infra failure, the build is marked red 🔴.
      These can manifest in various ways. In general, if the build logs don't
      fall in the categories above, there's a high chance it's an infrastructure
      failure. Common patterns usually have several `java` or `hudson` messages.
    * **Actions**: Infrastructure failures haven't been identified.

## Triggering CI

CI is usually triggered automatically when code is pushed or a pull request is
opened. But often it's necessary to re-trigger CI without making changes to the
code if:

* There were infrastructure failures in a previous build.
* There are fixes upstream, on infra or dependencies, that may make a new build pass.

The following methods can be used by maintainers to re-trigger builds:

* **Jenkins**

>     * Make a comment starting with `@osrfbuild run tests` on the pull request

Reviewer 3: I thought it was `@osrf-jenkins`? Is there a difference between the two?
